src/rez/utils/pip.py

```diff
@@ -0,0 +1,705 @@
+            # pair of the prerelease phase (the string "a", "b", or "rc") and the prerelease number
+            # the following conversions (-->) take place:
+            # a --> a, alpha --> a, b --> b, beta --> b, c --> c, rc --> rc, pre --> rc, preview --> rc
+            phase, number = pkg_version.pre
+            rez_version += '.' + phase + str(number)
+
+        if pkg_version.is_postrelease:
+            # this attribute will be the postrelease number (an integer)
+            # the following conversions (-->) take place:
+            # post --> post, rev --> post, r --> post
+            rez_version += ".post" + str(pkg_version.post)
+
+        if pkg_version.is_devrelease:
+            # this attribute will be the development release number (an integer)
+            rez_version += ".dev" + str(pkg_version.dev)
+
+        if pkg_version.local:
+            # representation of the local version portion is any
+            # the following conversions (-->) take place:
+            # 1.0[+ubuntu-1] --> 1.0[-ubuntu.1]
+            rez_version += "-" + pkg_version.local
+
+    return rez_version
+
+
+def pip_specifier_to_rez_requirement(specifier):
+    """Convert PEP440 version specifier to rez equivalent.
+
+    See https://www.python.org/dev/peps/pep-0440/#version-specifiers
+
+    Note that version numbers in the specifier are converted to rez equivalents
+    at the same time. Thus a specifier like '<1.ALPHA2' becomes '<1.a2'.
+
+    Note that the conversion is not necessarily exact - there are cases in
+    PEP440 that have no equivalent in rez versioning. Most of these are
+    specifiers that involve pre/post releases, which don't exist in rez (or
+    rather, they do exist in the sense that '1.0.post1' is a valid rez version
+    number, but it has no special meaning).
+
+    Note also that the specifier is being converted into rez format, but in a
+    way that still expresses how _pip_ interprets the specifier. For example,
+    '==1' is a valid version range in rez, but '==1' has a different meaning to
+    pip than it does to rez ('1.0' matches '==1' in pip, but not in rez). This
+    is why '==1' is converted to '1+<1.1' in rez, rather than '==1'.
+
+    Example conversions:
+
+        |   PEP440    |     rez     |
+        |-------------|-------------|
+        | ==1         | 1+<1.1      |
+        | ==1.*       | 1           |
+        | >1          | 1.1+        |
+        | <1          | <1          |
+        | >=1         | 1+          |
+        | <=1         | <1.1        |
+        | ~=1.2       | 1.2+<2      |
+        | ~=1.2.3     | 1.2.3+<1.3  |
+        | !=1         | <1|1.1+     |
+        | !=1.2       | <1.2|1.2.1+ |
+        | !=1.*       | <1|2+       |
+        | !=1.2.*     | <1.2|1.3+   |
+
+    Args:
+        specifier (str): PEP440-compatible specifier string.
+
+    Returns:
+        `VersionRange`: Equivalent rez version range.
+    """
+    ranges = []
+    parts = specifier.split(',')
```

Comment:
If you passed the `SpecifierSet` as mentioned lower, you could simply do `list(specifier)`. `packaging` will take care of removing all spaces.

Also, by passing the specifier set object, I think we could benefit from the packaging API here. For example, `list` on a `SpecifierSet` returns specifiers, from which you can get the operator with `.operator` and the version with `.version`. From the version we can do `packaging.version.Version(v)` and then use the methods to know if it's a pre/post release for example. We might still have to do some `endswith` but at least we will do way less than we do right now.

Reply:
I'm getting to this soon - yeah I'll redo this using native SpecifierSet as it looks like that'll simplify things